stfl/help.stfl

```diff
@@ -1,4 +1,4 @@
-vbox
+!vbox
```

Comment:
This works when initially opening the dialog, but it looks like it leaves the qnainput with focus if it was ever opened (e.g. for searching or after pressing ':'.
It looks like this can be resolved by putting the exclamation mark in front of the `textview[helptext]` line (verified with stfl-0.24)

Reply:
Thanks! I confirm the problem (with stfl-0.22), but the suggested fix doesn't work for me. When I apply it, keybinds stop working just as described in #2351. I added some logging and it appears that after exiting the command line, we're focusing `qnainput` again. This happens because `FormAction::finished_qna()` calls `set_focus("feeds")`, i.e. it attempts to focus a widget with hard-coded name. If a widget doesn't exist, STFL seems to keep the old focus. I changed the call to focus `article` widget instead, and the bug in the itemview was fixed.

I guess it's a matter of telling each `FormAction` how its main widget is called, so it knows what to focus.

Unfortunately I ran out of time for today, but I'll try my guess on the weekend and request a new review if that works.

Reply:
Sounds good.
In case that doesn't work, I've created yet another alternative: https://github.com/newsboat/newsboat/pull/2365
That PR sidesteps all of the introduced focus issues by reverting the changes and applying an alternative, minimal solution.

I still prefer your (this) PR because that keeps more of the STFL code out of the C++ code.

Reply:
I implemented my guess and it works (on 0.22 at least); see b1028db4a9eaf3215b99c7b725555e6b2544fd24. Can you please test it too?

Reply:
Yes, I just tested it with STFL versions 0.24 and 0.22.

It works fine for all dialogs except for the ItemView and the Help dialogs (pressing `/` followed by `ESC` will keep the qnainput focussed).
When applying the following changes on top of this PR, it starts working for all dialogs on both of the tested STFL versions:
```
diff --git a/include/formaction.h b/include/formaction.h
index 1d5514c6d2bc..f7dbe8b43a5c 100644
--- a/include/formaction.h
+++ b/include/formaction.h
@@ -147,7 +147,6 @@ private:
 	Operation finish_operation;
 	History* qna_history;
 	std::shared_ptr<FormAction> parent_formaction;
-	std::string focus_before_qna;
 };
 
 } // namespace newsboat
diff --git a/src/formaction.cpp b/src/formaction.cpp
index e9bd2cd2dba9..9da4f89424e2 100644
--- a/src/formaction.cpp
+++ b/src/formaction.cpp
@@ -428,8 +428,7 @@ void FormAction::cancel_qna()
 	f.set("show_msg", "1");
 	f.set("msg", "");
 
-	f.set_focus(focus_before_qna);
-	focus_before_qna.clear();
+	f.set_focus(main_widget());
 
 	v->inside_qna(false);
 	v->inside_cmdline(false);
@@ -605,10 +604,6 @@ void FormAction::start_next_question()
 		f.set("show_qna_input", "1");
 		f.set("show_msg", "0");
 
-		if (focus_before_qna.empty()) {
-			focus_before_qna = f.get_focus();
-		}
-
 		f.set_focus("qnainput");
 
 		// Set position to 0 and back to ensure that the text is visible
@@ -628,8 +623,7 @@ void FormAction::start_next_question()
 		f.set("show_msg", "1");
 		f.set("msg", "");
 
-		f.set_focus(focus_before_qna);
-		focus_before_qna.clear();
+		f.set_focus(main_widget());
 
 		this->finished_qna(finish_operation);
 	}
```

Reply:
Thanks, those changes work great!

I wanted to finish this off by using `main_widget()` in place of `!` in STFL, and now focus is broken again :roll_eyes: I'll try to figure out why, tomorrow.

Reply:
Well, whatever, I couldn't figure it out, and I'd rather push what I already have than polish it forever. So please see my latest two commits, and let's get this merged for 2.31 :)

Reply:
Sounds good, thank you for pushing this to completion :)